includes/class-wc-stripe-api.php

```diff
@@ -250,6 +278,84 @@ public static function retrieve( $api ) {
 		return json_decode( $response['body'] );
 	}
 
+	/**
+	 * Checks if the Stripe API for the current mode (i.e. test or live) is rate limited.
+	 *
+	 * @return bool True if the Stripe API is rate limited, false otherwise.
+	 */
+	public static function is_stripe_api_rate_limited() {
+		$rate_limit_option_key = WC_Stripe_Mode::is_test() ? self::TEST_MODE_STRIPE_API_RATE_LIMIT_OPTION_KEY : self::LIVE_MODE_STRIPE_API_RATE_LIMIT_OPTION_KEY;
+
+		$rate_limit_expiration = get_option( $rate_limit_option_key );
+		if ( ! $rate_limit_expiration ) {
+			return false;
+		}
+
+		$now = time();
+		if ( $now > $rate_limit_expiration ) {
+			delete_option( $rate_limit_option_key );
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Helper function to check error responses from Stripe and ensure we prevent unnecessary API calls,
+	 * primarily in cases where we have been rate limited or we don't have valid keys..
+	 *
+	 * @param array|WP_Error $response The response from the Stripe API.
+	 * @return void
+	 */
+	protected static function check_stripe_api_error_response( $response ) {
+		// If we don't have an array for $response, return early, as we won't have an HTTP status code.
+		if ( ! is_array( $response ) ) {
+			return;
+		}
+
+		// We specifically want to check $response['response']['code']. If it's not present, return early.
+		if (
+			! isset( $response['response'] )
+			|| ! is_array( $response['response'] )
+			|| ! isset( $response['response']['code'] )
+		) {
+			return;
+		}
+
+		$status_code = $response['response']['code'];
+
+		if ( 429 === $status_code ) {
```

Comment:
Based on https://docs.stripe.com/rate-limits#rate-limited-requests, it is better to check for the header. It is possible that a 429 may not be related to rate-limiting. Both work fine, but checking the header would be more consistent.